### DISCUSSION TITLE
LUA features

### DISCUSSION THREAD

**Discussion by omegahim** [YimMenu/YimMenu#1389] [General]
With the new Lua Scripts option, where can the scripts even be found and downloaded?

**Comment by Yimura** [comment]
Nowhere currently, the Lua feature is being rolled out as a beta, neither is there any documentation available so you'll have to look into the source code to see what is available.

However people can for now share any Lua scripts they have here.

**Reply by Fhonixx** [reply]
How do i know if i can use Lua scripts??

**Reply by Tayipc38** [reply]
Youll have to be patient untill some documentation or sample scripts get released.

**Reply by Fhonixx** [reply]
ohh ok ty for telling me that (:

**Comment by tupoy-ya** [comment]
#### Didn't test any of these on stock YimMenu.

### Phone Animation from single player
May not be ideal, just a port from TupoyeMenu.
```lua
script.register_looped("phone_anim", function()
	local self_ped = PLAYER.PLAYER_PED_ID() --TODO: Make a self library.
	if not PED.GET_PED_CONFIG_FLAG(self_ped, 244, false) and not ENTITY.IS_ENTITY_DEAD(self_ped, false) then
		PED.SET_PED_CONFIG_FLAG(self_ped, 244, false)
		PED.SET_PED_CONFIG_FLAG(self_ped, 243, false)
		PED.SET_PED_CONFIG_FLAG(self_ped, 242, false)
	end
end)
```

### Basic `self` library
Put into `Grand Theft Auto V/lua/self_lib.lua`, as the appdata `scripts` directory doesn't seem to be in the search path.
Usage in the [Floats not working bug](#floats-not-working-bug).
```lua
self = {}

function self.id()
	return PLAYER.PLAYER_ID()
end

function self.ped()
	return PLAYER.PLAYER_PED_ID()
end

function self.pos()
	return ENTITY.GET_ENTITY_COORDS(self.ped(), false)
end

function self.veh()
	if PED.IS_PED_IN_ANY_VEHICLE(self.ped(), 0) then
		return PED.GET_VEHICLE_PED_IS_IN(self.ped(), false);
	end

	return 0
end

return self
```

### Floats not working bug
Any natives with floats as parameters seem to always default to 0 no matter what value you set.
For example looping:
```lua
self = require "self_lib"

script.register_looped("loop_name", function()
	PED.SET_PED_COORDS_KEEP_VEHICLE(self.ped(), 100.0, 10, 20.2);
end)
```
Will just teleport you at 0 0 0.

Could be a TupoyeMenu bug, but i didn't change anything related to floats in there.

###### TODO: GUI always crashing bug.

###### EDIT: The above statement is false, i just used it wrong.
###### Anyway it should not be crashing if used incorrectly.

**Reply by Adnanou07** [reply]
Yo where to put the lua scripts, can't find the search path. 
I ve got some good lua scripts from other mod menus, tried to upload it to yimmenu but didnt find where to put it

**Reply by tupoy-ya** [reply]
> Yo where to put the lua scripts, can't find the search path

All .lua scripts in this directory will be executed: `drive_c/users/steamuser/AppData/Roaming/BigBaseV2/scripts/`

This is a search path for includes: `Grand Theft Auto V/lua/`

**Reply by Adnanou07** [reply]
Ok thanks so much 🙏

**Reply by Yimura** [reply]
Kiddion modest menu scripts don’t work for YimMenu btw

**Reply by wangzixuank** [reply]
how to teleport?
PED.SET_PED_COORDS_KEEP_VEHICLE(PLAYER.PLAYER_PED_ID(), 100.0, 10, 20.2);
not working

**Reply by tupoy-ya** [reply]
> how to teleport? 
> ```lua
> PED.SET_PED_COORDS_KEEP_VEHICLE(PLAYER.PLAYER_PED_ID(), 100.0, 10, 20.2);
> ```
> not working

From my testing anything with float as an argument will just treat it as zero, so i guess teleport is not possible.

I could be wrong, but i have not found a way to use float values in YimMenu lua.

**Comment by Tayipc38** [comment]
> ### Basic `self` library
> 
> ```lua
> self = {}
> 
> function self.id()
> 	return PLAYER.PLAYER_ID()
> end
> 
> function self.ped()
> 	return PLAYER.PLAYER_PED_ID()
> end
> 
> function self.pos()
> 	return ENTITY.GET_ENTITY_COORDS(self.ped(), false)
> end
> 
> function self.veh()
> 	if PED.IS_PED_IN_ANY_VEHICLE(self.ped(), 0) then
> 		return PED.GET_VEHICLE_PED_IS_IN(self.ped(), false);
> 	end
> 
> 	return 0
> end
> 
> return self
> ```

I quite like the idea of libraries. It would be great if we started at least a little documentation on the lua (i know its beta).

**Comment by sch-lda** [comment]
Does anyone know how to add new tabs and buttons? I've tried the code below and this lua was successfully loaded, with "GUI Tabs Registered" displayed as 1. However, I couldn't find any added buttons or new tabs.Is this a program bug or did I use the wrong syntax?

local tab_test = gui.get_tab("tab_test")
local test_btn = tab_test.add_button("btn", function()
    gui.show_message("title", "msg")
end)

**Reply by wangzixuank** [reply]
there is no tab name "tab_test"

**Reply by sch-lda** [reply]
Is it not possible to create new tabs , and can new buttons only be added to existing tabs instead?
I also tried MOBILE, but I did not see any new buttons on the mobile page.

local tab_test = gui.get_tab("MOBILE")
local test_btn = tab_test.add_button("btn", function()
gui.show_message("title", "msg")
end)

**Reply by wangzixuank** [reply]
check src\services\gui\gui_service.hpp

**Reply by tupoy-ya** [reply]
You need to use `:` instead of `.` on `tab_test`

Here is some test GUI Lua that i tested works.

`get_tab` uses tabs translation label.
```lua
gui.get_tab("GUI_TAB_ESP"):add_text("Text")
```

You should be able to store it in a variable and call multiple methods on the tab.

**Comment by pierrelasse** [comment]
It would be great if someone could make a script that lets you make money, since moddest is currently outdated and yimmenu has no money making related stuff.

Also a bounty remover would be great.

**Reply by sch-lda** [reply]
remove bounty: globals.set_int(1+2359296+5150+13,2880000)   

I'm not sure what methods you are accustomed to using to earn money, but I believe that utilizing in-game missions and tasks within the game itself is relatively safer. You can find  Heist Editor and a method to instant complete the final chapter of the Perico Heist on unknowncheats forum.
perico instant pass:
locals.set_int("fm_mission_controller_2020","46829","50")
locals.set_int("fm_mission_controller_2020","45450","9")

**Reply by pierrelasse** [reply]
Thanks for the response and help.

**Comment by sch-lda** [comment]


**Reply by who-care123** [reply]
Set the entity as mission entity before deletion

**Comment by xiaoxiao921** [comment]
Finally documented the whole lua api we got so far.

Until we put the pages on yim.gta.menu website you can check the available documentation with these two links:

All the lua tables so far:
https://github.com/YimMenu/YimMenu/tree/0685ab1d155547cc28d6512c4126e98f795fe987/src/lua/docs/tables

All the lua classes so far:
https://github.com/YimMenu/YimMenu/tree/0685ab1d155547cc28d6512c4126e98f795fe987/src/lua/docs/classes

I'm thinking of a solution for all the `commands` we have, so that they are also in the lua api doc too

**Comment by iDominiqe** [comment]
Wanted to ask about the attitude and support of lua along with xml files and .js if they work together or separately. I saw a lot of scripts on the Internet that support different formats for creating additional map modifications, animated objects that personalize players' clothes and vehicles. I'm a little dumb, and wanted to ask experienced people, thanks in advance.
https://www.guilded.gg/stand/groups/x3ZgB10D/channels/bae59c11-4c7a-4a77-9738-d3876a47c786/docs/302105
https://www.gta5-mods.com/scripts/tags/lua
https://www.kiddionsmodestmenu.com/modest-scripts
not advertising 🙂

**Comment by sch-lda** [comment]
https://github.com/YimMenu/YimMenu/issues/1777
Anyone knows how to get all peds/objects/vehicles as handles using natives？Thanks in advance.

**Reply by sch-lda** [reply]
Thank you for your explanation.
